extensions/widgets/switchbutton/switchbutton.lcb

```diff
@@ -17,199 +17,709 @@ along with LiveCode.  If not see <http://www.gnu.org/licenses/>.  */
+	if mIsPressed then
+		variable tXMousePos as Real
+		variable tChange as Real
+		variable tMovedRight as Boolean
+
+		put true into mMouseHasMoved
+
+		put the x of the mouse position into tXMousePos
+		put tXMousePos - mXClick into tChange
+		if tChange > 0 then
+			put true into tMovedRight
+		else
+			put false into tMovedRight
+		end if
+
+		if tMovedRight then
+			put true into mSwitchIsOn
+		else
+			put false into mSwitchIsOn
+		end if
+
 		redraw all
 	end if
+end handler
 
-	post "mouseUp" with [mIsChecked]
+public handler OnMouseDown() returns nothing
+	if the click position is within clickableRect() then
+		put the x of the click position into mXClick
+		put true into mIsPressed
+		redraw all
+	end if
 end handler
 
-private handler getPath(in pShape as String) returns Path
-	if pShape is "box" then
-		return rounded rectangle path of rectangle [(my width)/2 - (my height)/2, (my height)/4, (my width)/2 + (my height)/2, 3*(my height)/4] with radius (my height)/4
-	else if pShape is "left circle" then
-		return circle path centered at point [(my width)/2 - (my height)/4, (my height)/2] with radius (my height)/4
-	else if pShape is "right circle" then
-		return circle path centered at point [(my width)/2 + (my height)/4, (my height)/2] with radius (my height)/4
-	else if pShape is "ellipse" then
-		return ellipse path centered at point [(my width)/2, (my height)/2] with radii [(my height)/8, (my height)/4]
-	else if pShape is "left rect" then
-		return rectangle path of rectangle [(my width)/2 - (my height)/4, (my height)/4, (my width)/2, 3*(my height)/4]
-	else if pShape is "right rect" then
-		return rectangle path of rectangle [(my width)/2, (my height)/4, (my width)/2 + (my height)/4, 3*(my height)/4]
+public handler OnMouseUp() returns nothing
+	if mMouseHasMoved then
+		post "mouseUp" with [mSwitchIsOn]
+	else
+		if the click position is within clickableRect() then
+			setSwitch(not(mSwitchIsOn))
+			post "mouseUp" with [mSwitchIsOn]
+		end if
 	end if
+
+	put false into mIsPressed
+	redraw all
 end handler
 
-private handler getRect(in pObject as String) returns Rectangle
-	if pObject is "bounding box" then
-		return rectangle [(my width)/2 - 3*(my height)/4, (my height)/4, (my width)/2 + 3*(my height)/4, 3*(my height)/4]
+--------------------------------------------------------------------------------
+--
+--		Setting Properties
+--
+--------------------------------------------------------------------------------
+
+constant kWidgetThemeIos is "iOS"
+constant kWidgetThemeAndroidDark is "Android(Dark)"
+constant kWidgetThemeAndroidLight is "Android(Light)"
+
+private handler setWidgetTheme(in pWidgetTheme as String) returns nothing
+
+	variable tThemeList as List
+	put [kWidgetThemeIos, kWidgetThemeAndroidDark, kWidgetThemeAndroidLight] into tThemeList
+
+	if not(pWidgetTheme is in tThemeList) then
+		throw "invalid widget theme"
 	end if
+
+	if pWidgetTheme is mWidgetTheme then
+		return
+	end if
```

Comment:
Could you put this bit at the top of the handler too, so it doesn't do anything if the theme doesn't change